llvm/test/CodeGen/AArch64/sve-cmp-folds.ll

```diff
@@ -114,9 +114,12 @@ define i1 @foo_last(<vscale x 4 x float> %a, <vscale x 4 x float> %b) {
 ; CHECK-LABEL: foo_last:
 ; CHECK:       // %bb.0:
 ; CHECK-NEXT:    ptrue p0.s
-; CHECK-NEXT:    fcmeq p1.s, p0/z, z0.s, z1.s
-; CHECK-NEXT:    ptest p0, p1.b
-; CHECK-NEXT:    cset w0, lo
+; CHECK-NEXT:    mov x8, #-1 // =0xffffffffffffffff
```

Comment:
I don't speak AArch64 but this looks like an unfortunate regression, at least in terms of number of instructions.

Reply:
Agreed, ping @sdesmalen-arm, just to make you guys aware.

Reply:
For the following IR:
```
define i64 @foo_last() {
  %vscale = call i64 @llvm.vscale.i64()
  %shl2 = shl nuw nsw i64 %vscale, 2
  %idx = add nuw nsw i64 %shl2, -1
  ret i64 %idx
}

declare i64 @llvm.vscale.i64()
```
With the current patch this is incorrectly simplified to:
```
t9: ch,glue = CopyToReg t0, Register:i64 $x0, Constant:i64<-1>
```

I don't think there's anything AArch64-specific going on there.

Reply:
That simplification looks correct to me. The only way `add -1` can the nuw is if the other operand is zero, so the result is -1. This probably wasn't supposed to have a nuw flag?

Reply:
Doh :) Yes I see it now. It should have been `sub nuw nsw i64 shl2, 1` to avoid any wrapping (or indeed remove the nuw flag). I'll have a look to see if we generate this pattern with nuw flag anywhere. Thanks!

Reply:
`vscale` can never be zero though (it should have a minimum value of 1). Should it be returning `poison` in that case?